facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProvider.java

```diff
@@ -0,0 +1,174 @@
+        super.validateBeforeSaving(restEntity);
+        validateDoesNotConflict(restEntity);
+    }
+
+    private void validateDoesNotConflict(ProductMilestone restEntity)
+            throws ConflictedEntryException, InvalidEntityException {
+        ValidationBuilder.validateObject(restEntity, WhenUpdating.class).validateConflict(() -> {
+            org.jboss.pnc.model.ProductMilestone milestoneFromDB = repository.queryByPredicates(
+                    withProductVersionIdAndVersion(restEntity.getProductVersion().getId(), restEntity.getVersion())
+            );
+
+            // don't validate against myself
+            if (milestoneFromDB != null && !milestoneFromDB.getId().equals(restEntity.getId())) {
+                return new ConflictedEntryValidator.ConflictedEntryValidationError(
+                        milestoneFromDB.getId(),
+                        org.jboss.pnc.model.ProductMilestone.class,
+                        "Product milestone with the same product version and version already exists");
+            }
+            return null;
+        });
+    }
+
+    public void closeMilestone(int id, ProductMilestone restEntity) {
+        closeMilestone(id, restEntity, "");
+    }
+
+    public void closeMilestone(int id, ProductMilestone restEntity, String accessToken) {
+
+        org.jboss.pnc.model.ProductMilestone milestoneInDb = repository.queryById(id);
+
+        if (milestoneInDb.getEndDate() != null) {
+
+            log.info("Milestone is already closed: no more modifications allowed");
+            throw new RepositoryViolationException("Milestone is already closed! No more modifications allowed");
+        } else {
+            // save download url if specified
+            if (restEntity.getDownloadUrl() != null) {
+
+                milestoneInDb.setDownloadUrl(restEntity.getDownloadUrl());
+                repository.save(milestoneInDb);
+            }
+
+            if(releaseManager.noReleaseInProgress(milestoneInDb)) {
+                log.debug("Milestone end date set and no release in progress, will start release");
+                releaseManager.startRelease(milestoneInDb, accessToken);
+            }
+        }
+    }
+
+    public void cancelMilestoneCloseProcess(Integer id)
+            throws RepositoryViolationException, EmptyEntityException {
+
+        org.jboss.pnc.model.ProductMilestone milestoneInDb = repository.queryById(id);
+
+        // If we want to close a milestone, make sure it's not already released (by checking end date)
+        // and there are no release in progress
+        if (milestoneInDb.getEndDate() != null) {
+
+            log.info("Milestone is already closed.");
+            throw new RepositoryViolationException("Milestone is already closed!");
+
+        } else {
+
+            if(releaseManager.noReleaseInProgress(milestoneInDb)) {
+
+                log.debug("Milestone end date set and no release in progress, will start release");
```

Comment:
I don't get the meaning od this.

Reply:
Edited the log slightly to say:

```
Milestone's 'end date' set; no release of the milestone in progress: will start release
```

Reply:
Yeah, but you are not starting anything, you are throwing an exception... I don't understant what happens there.

Reply:
It's 'log.debug', not 'throw new Exception' :)

Reply:
Eyy I see it now. Fixed! :)